### PR TITLE
OPSEXP-1802 Pre-Commit Hooks Best Practices in Build Tools

### PR DESCRIPTION
**Pull request by upandey123**

### CODE REVIEW

**Comment by upandey123** [comment]
@gionn I have renamed the file to .md now. What else were you explaining me to check in order to avoid white spaces issue ?

File: docs/pre-commit-hooks.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+Git hooks are scripts that run automatically every time a particular event occurs in a Git repository. They let you customize Git’s internal behavior and trigger customizable actions at key points in the development life cycle. Git hooks are a useful way to run scripts when an action such as a commit, merge, or push occurs.
```
Comment (gionn) [inline docs/pre-commit-hooks.md:1]:
```suggestion
# Pre-commit hooks

Git hooks are scripts that run automatically every time a particular event occurs in a Git repository. They let you customize Git’s internal behavior and trigger customizable actions at key points in the development life cycle. Git hooks are a useful way to run scripts when an action such as a commit, merge, or push occurs.
```

File: docs/pre-commit-hooks.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+To prevent the developer from committing bad code and running CI builds that would break the environment, we decided to use the pre-commit library to manage hooks and run some basic checks on the code. All the information about the pre-commit library can be found in the official documentation.
+
+The goal of this guide is to provide a quick starting point to the developer and explain how we are using it on the single pipeline repository and how we can configure it for other projects
+
+pre-commit runs locally while performing a commit, but with a small effort, it is possible to run it in CI too, during the CI build.
+
+Below are the few Hooks and their description,
+
+
```
Comment (gionn) [inline docs/pre-commit-hooks.md:11]:
```suggestion
```

File: docs/pre-commit-hooks.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+1) Helm-deps (Helm Dependency Update):-
```
Comment (gionn) [inline docs/pre-commit-hooks.md:13]:
```suggestion
## Helm-deps (Helm Dependency Update)
```

use headings for each component

File: docs/pre-commit-hooks.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+Manage the dependencies of a chart.
+
+# Chart.yaml
+dependencies:
+- name: nginx
+  version: "1.2.3"
+  repository: "https://example.com/charts"
+- name: memcached
+  version: "3.2.1"
+  repository: "https://another.example.com/charts"
```
Comment (gionn) [inline docs/pre-commit-hooks.md:24]: wrap multiline code with backticks (```)

**Review by gionn — COMMENTED** [review]
added a few inline comments, for a brief markdown introduction take a look at https://github.com/jorditorresBCN/Quick-Start/blob/master/Quick-Start-Markdown.md

File: docs/pre-commit-hooks.md (hunk 1 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md:3]: fixed

File: docs/pre-commit-hooks.md (hunk 3 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md]: fixed

File: docs/pre-commit-hooks.md (hunk 4 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md]: fixed

File: docs/pre-commit-hooks.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# Pre-commit Hooks
+
+Git hooks are scripts that run automatically every time a particular event occurs in a Git repository. They let you customize Git’s internal behavior and trigger customizable actions at key points in the development life cycle. Git hooks are a useful way to run scripts when an action such as a commit, merge, or push occurs.
+
+To prevent the developer from committing bad code and running CI builds that would break the environment, we decided to use the pre-commit library to manage hooks and run some basic checks on the code. All the information about the pre-commit library can be found in the official documentation.
+
+The goal of this guide is to provide a quick starting point to the developer and explain how we are using it on the single pipeline repository and how we can configure it for other projects
```
Comment (gionn) [inline docs/pre-commit-hooks.md:7]: on the single pipeline?

File: docs/pre-commit-hooks.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+pre-commit runs locally while performing a commit, but with a small effort, it is possible to run it in CI too, during the CI build.
+
+Below are the few Hooks and their description,
```
Comment (gionn) [inline docs/pre-commit-hooks.md:11]:
```suggestion
Below are the hooks available in this repository and defined in [.pre-commit-hooks.yaml](../.pre-commit-hooks.yaml)
```

File: docs/pre-commit-hooks.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+- name: memcached
+  version: "3.2.1"
+  repository: "https://another.example.com/charts"
+```
+
+The 'name' should be the name of a chart, where that name must match the name in that chart's 'Chart.yaml' file.
+
+The 'version' field should contain a semantic version or version range.
+
+The 'repository' URL should point to a Chart Repository. Helm expects that by appending '/index.yaml' to the URL, it should be able to retrieve the chart repository's index. Note: 'repository' can be an alias. The alias must start with 'alias:' or '@'.
```
Comment (gionn) [inline docs/pre-commit-hooks.md:31]:
```suggestion
Validate the dependencies of a chart.

Add to your `.pre-commit-config.yaml`:


  - repo: https://github.com/Alfresco/alfresco-build-tools
    rev: v1.22.0
    hooks:
      - id: helm-deps

```

Comment (gionn) [reply docs/pre-commit-hooks.md:31]: provides a similar structure for all the hooks defined in https://github.com/Alfresco/alfresco-build-tools/blob/master/.pre-commit-hooks.yaml

**Review by gionn — CHANGES_REQUESTED** [review]

**Comment by upandey123** [comment]
Hi @alxgomz as Gio is out, could you please review and approve accordingly?

File: docs/pre-commit-hooks.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        with:
+          version: $HELM_VERSION
+        id: install
+
+      - uses: cpanato/kubepug-installer@v1.0.0
+
+      - name: Run Kubepug with your Helm Charts Repository
+        run: |
+          find charts -mindepth 1 -maxdepth 1 -type d | xargs -t -n1 -I% /bin/bash -c 'helm template % --api-versions ${K8S_TARGET_VERSION} | kubepug --error-on-deprecated --error-on-deleted --k8s-version ${K8S_TARGET_VERSION} --input-file /dev/stdin'
+```
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:82]: Why is this example using github actions integration while the page is about pre-commit hooks?
use the pre-commit action instead

File: docs/pre-commit-hooks.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      - id: plantuml
+```
+
+During the commit of any PlantUML file (file with extension: .puml), if we
+
+add or update a PlantUML file, a PlantUML image will be generated based on the file and commit
+
+delete a PlantUML file, existing PlantUML image will be deleted together in the commit
+
+Open your pre-commit file and add the following code to run plantuml.pre-commit.sh
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:100]:
```suggestion
When introducing or updating a PlantUML file (with extension `.puml`), an image will
be generated based on the file. Also removing a PlantUML file from the git index will trigger deletion of  existing PlantUML image.
Open your pre-commit file and add the following code to run plantuml.pre-commit.sh
```

File: docs/pre-commit-hooks.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+```
+
+## Plantuml
+
+Add to your `.pre-commit-config.yaml`:
+
+```  - repo: https://github.com/Alfresco/alfresco-build-tools
+    rev: v1.22.0
+    hooks:
+      - id: plantuml
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:91]: there is no such hook in the repo: https://github.com/Alfresco/alfresco-build-tools/blob/master/.pre-commit-hooks.yaml
Choose one that does exists

File: docs/pre-commit-hooks.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+KubePug is a kubectl plugin checking for deprecated Kubernetes clusters or deprecated versions of Kubernetes manifests.
+
+Validate this hook.
+
+Add to your `.pre-commit-config.yaml`:
+
+```  - repo: https://github.com/Alfresco/alfresco-build-tools
+    rev: v1.22.0
+    hooks:
+      - id: kubepug
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:54]: there is no such hook in the repo: https://github.com/Alfresco/alfresco-build-tools/blob/master/.pre-commit-hooks.yaml
Choose one that does exists

File: docs/pre-commit-hooks.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+delete a PlantUML file, existing PlantUML image will be deleted together in the commit
+
+Open your pre-commit file and add the following code to run plantuml.pre-commit.sh
+
+```#!/bin/bash
+...
+
+"$(git rev-parse --git-dir)/hooks/plantuml.pre-commit.sh"
+```
+
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:107]: Can you clarify why this is needed? and where it needs to go

File: docs/pre-commit-hooks.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+## Checkov
+
+To use Checkov with pre-commit, just add the following to your local repo’s .pre-commit-config.yaml file:
+
+```- repo: https://github.com/bridgecrewio/checkov.git
+  rev: '' # change to tag or sha
+  hooks:
+    - id: checkov
+```
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:116]: Why do you document using an external hook instead of the one used in this repo?

**Review by alxgomz — APPROVED** [review]

**Review by alxgomz — CHANGES_REQUESTED** [review]
please fix issues mentionned

File: docs/pre-commit-hooks.md (hunk 12 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md:107]: I took the reference of this but it seems pretty long so deleted that.

https://github.com/weikangchia/pre-commit-hooks-plantuml

File: docs/pre-commit-hooks.md (hunk 13 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md:116]: fixed

File: docs/pre-commit-hooks.md (hunk 11 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md:54]: done

File: docs/pre-commit-hooks.md (hunk 10 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md:91]: done

File: docs/pre-commit-hooks.md (hunk 8 above)
Comment (upandey123) [reply docs/pre-commit-hooks.md:82]: done

**Comment by upandey123** [comment]
@alxgomz please review again

**Review by alxgomz — CHANGES_REQUESTED** [review]

**Review by alxgomz — CHANGES_REQUESTED** [review]
your markdown code blocks for yaml are malformated.
Use ```yaml and start writing content on a new line

File: docs/pre-commit-hooks.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    hooks:
+      - id: helm-lint
+```
+
+Helm lint examine a chart for potential issues and verify the chart is well-formed. Use --strict to return an error for the step if a chart isn't formatted properly.
+
+```yaml
+  - name: Lint Charts
+    run: helm lint --strict $HELM_WORKING_DIRECTORY
+```
```
Comment (alxgomz) [inline docs/pre-commit-hooks.md:44]: That doesn't look applicable to our hook

**Review by alxgomz — CHANGES_REQUESTED** [review]

**Comment by upandey123** [comment]
As Gio is on vacation so removed his review.

**Pull request merged**